### PR TITLE
Test-driven-development

### PR DESCRIPTION
**Pull request by keebot-png**
## The following has been completed in line with project requirements:
- [x] Set up Spec folder with multiple unit test files.
- [x] Wrote unit tests for the `Solver` class,  `fizz buzz`, `reverse` and `factorial` methods.
- [x] Fix linter errors.
- [x] Correct Gitflow.
- [x] Update Readme file.

### CODE REVIEW

File: spec/solver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    it 'reverse the input string' do
+      expect(@solver.reverse('keenan')).to eq 'naneek'
+    end
+  end
+
+  describe '#fizzbuzz' do
+    it 'returns constructed string' do
+      expect(@solver.fizzbuzz(15)).to eq '1, 2, fizz, 4, buzz, fizz, 7, 8, fizz, buzz, 11, fizz, 13, 14, fizzbuzz'
+    end
+  end
```
Comment (Ghiftee) [inline spec/solver_spec.rb]: - [x] Good job writing the tests, however, according to the requirements below, kindly add the following tests for factorial method
- Negative value which raises an exception as an expectation
- Special case - when input is 0, the result should equal 1

- [x] For the fizzbuzz method, kindly add the following tests
```ruby
expect(resolver.fizzbuzz(3)).to eql('fizz')  # N is divisible by 3
    ...
    expect(resolver.fizzbuzz(5)).to eql('buzz')  # N is divisible by 5
    ...
    expect(resolver.fizzbuzz(15)).to eql('fizzbuzz') # N is divisible by 3 and 5
    ...
    expect(resolver.fizzbuzz(7)).to eql('7')  # N is not divisible by 3 or 5
```

Screenshot 2023-02-10 at 11 04 01 (2)

**Review by Ghiftee — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️♻️
Hello @keebot-png and @yonas44 👋🏽 

Good job working on the project so far! 
To highlight 👍🏽 
✔️ You used TDD
✔️ Gitflow is used
✔️ No linter errors

However, there are some issues that you need to work on to go to the next project but you are almost there!



## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: spec/solver_spec.rb (hunk 1 above)
Comment (yonas44) [reply spec/solver_spec.rb]: Okay thank you for your feedback🙏

File: spec/solver_spec.rb [hunk 2]
```diff
@@ -20,7 +22,7 @@
 
   describe '#fizzbuzz' do
     it 'returns constructed string' do
-      expect(@solver.fizzbuzz(15)).to eq '1, 2, fizz, 4, buzz, fizz, 7, 8, fizz, buzz, 11, fizz, 13, 14, fizzbuzz'
+      expect(@solver.fizzbuzz(15)).to eq 'fizzbuzz'
```
Comment (pinkmoon25) [inline spec/solver_spec.rb:25]: - [x] As requested by the previous reviewer https://github.com/keebot-png/TDD-project/pull/1#discussion_r1102551406 please update and add the below test cases: 
 ```
      expect(@solver.fizzbuzz(3)).to eq 'fizz'
      expect(@solver.fizzbuzz(5)).to eq 'buzz'
      expect(@solver.fizzbuzz(7)).to eq '7'
```

File: solver.rb [hunk 3]
```diff
@@ -10,20 +11,14 @@ def reverse(str)
...
-    end
-    str
+    if (num % 15).zero?
+        return 'fizzbuzz'
+    elsif (num % 3).zero?
+        return 'fizz'
+    elsif (num % 5).zero?
+        return 'buzz'
+    else
+        return num       i
```
Comment (pinkmoon25) [inline solver.rb]: - [x] please notice that the value `i` is undefined here, and you can use string interpolation to return value of num like `"#{num}"`

**Review by pinkmoon25 — CHANGES_REQUESTED** [review]
Hi @keebot-png @yonas44 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



 ## Highlights
gitflow :heavy_check_mark: 
Tdd :heavy_check_mark: 
linters :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/solver_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    it 'reverse the input string' do
+      expect(@solver.reverse('keenan')).to eq 'naneek'
+    end
+  end
+
+  describe '#fizzbuzz' do
+    it 'returns constructed string' do
+      expect(@solver.fizzbuzz(15)).to eq 'fizzbuzz'
+    end
+  end
```
Comment (tresorsawasawa) [inline spec/solver_spec.rb]: - [ ] As mentioned by the previous reviewer, kindly ensure that the `#fizzbuzz` method has also tests that handle the **[following aspects](https://github.com/keebot-png/TDD-project/pull/1#discussion_r1102551406)**:
  * Number is divisible by 3
  * Number is divisible by 5
  * Number is divisible by 3 and 5
  * Number is not divisible by 3
  * etc

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUEST CHANGES :recycle::recycle: 

Hi @yonas44 and @keebot-png 
Great work on making the changes requested by the previous reviewer  👏👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost-there

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Your design looks good :100: :heavy_check_mark: 
- Good PR description :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by tresorsawasawa — APPROVED** [review]
# STATUS: APPROVE :heavy_check_mark: :heavy_check_mark:  :champagne: 

Hi @keebot-png and @yonas44 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 :clap: :clap: :100: 

nodding-yes

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
- Your README file is professional :heavy_check_mark: 

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**